pkg/controllers/nodeclass/validation.go

```diff
@@ -324,19 +339,98 @@ func mockLaunchTemplateConfig() []ec2types.FleetLaunchTemplateConfigRequest {
+	// Issue: https://github.com/aws/karpenter-provider-aws/issues/7928
+	amiMap := amifamily.MapToInstanceTypes(v.getInstanceTypesForNodeClass(ctx, nodeClass), nodeClass.Status.AMIs)
+	var instanceTypes []*cloudprovider.InstanceType
+	for _, ami := range nodeClass.Status.AMIs {
+		if len(amiMap[ami.ID]) == 0 {
+			continue
+		}
+		amiRequirements := scheduling.NewNodeSelectorRequirements(ami.Requirements...)
+		if amiRequirements.IsCompatible(amifamily.VariantStandard.Requirements()) {
+			instanceTypes = append(instanceTypes, amiMap[ami.ID]...)
+		}
+	}
+	if len(instanceTypes) == 0 && len(amiMap) != 0 {
+		instanceTypes = lo.Flatten(lo.Values(amiMap))
+	}
+	// If there weren't any matching instance types, we should fallback to some defaults. There's an instance type included
+	// for both x86_64 and arm64 architectures, ensuring that there will be a matching AMI.
+	if len(instanceTypes) == 0 {
+		instanceTypes = []*cloudprovider.InstanceType{
+			{
+				Name: string(ec2types.InstanceTypeM5Large),
+				Requirements: scheduling.NewRequirements(append(
+					lo.Values(amifamily.VariantStandard.Requirements()),
+					scheduling.NewRequirement(corev1.LabelArchStable, corev1.NodeSelectorOpIn, karpv1.ArchitectureAmd64),
+				)...),
+			},
+			{
+				Name: string(ec2types.InstanceTypeM6gLarge),
+				Requirements: scheduling.NewRequirements(append(
+					lo.Values(amifamily.VariantStandard.Requirements()),
+					scheduling.NewRequirement(corev1.LabelArchStable, corev1.NodeSelectorOpIn, karpv1.ArchitectureArm64),
+				)...),
+			},
+		}
+	}
+	opts, err := v.amiResolver.Resolve(nodeClass, nodeClaim, instanceTypes, karpv1.CapacityTypeOnDemand, amiOptions)
+	if err != nil {
+		return nil, err
+	}
+	return opts[0], nil
+}
+
+// getInstanceTypesForNodeClass returns the set of instances which could be launched using this NodeClass based on the
+// requirements of linked NodePools. If no NodePools exist for the given NodeClass, this function returns two default
+// instance types (one x86_64 and one arm64).
+func (v *Validation) getInstanceTypesForNodeClass(ctx context.Context, nodeClass *v1.EC2NodeClass) []*cloudprovider.InstanceType {
+	instanceTypes, err := v.cloudProvider.GetInstanceTypes(ctx, &karpv1.NodePool{Spec: karpv1.NodePoolSpec{Template: karpv1.NodeClaimTemplate{Spec: karpv1.NodeClaimTemplateSpec{
+		NodeClassRef: &karpv1.NodeClassReference{
+			Group: object.GVK(nodeClass).Group,
+			Kind:  object.GVK(nodeClass).Kind,
+			Name:  nodeClass.Name,
 		},
-	}, karpv1.CapacityTypeOnDemand, amiOptions)
+	}}}})
+	if err != nil {
+		return nil
+	}
+
+	nodePools, err := nodepool.ListManaged(ctx, v.kubeClient, v.cloudProvider, nodepool.ForNodeClass(nodeClass))
+	if err != nil {
+		return nil
+	}
+	var compatibleInstanceTypes []*cloudprovider.InstanceType
+	names := sets.New[string]()
+	for _, np := range nodePools {
+		reqs := scheduling.NewNodeSelectorRequirementsWithMinValues(np.Spec.Template.Spec.Requirements...)
```

Comment:
Doesn't it also need to consider labels (probably less likely though)?

Reply:
Oh, I think you're right. You could do something like this and it would technically act as a requirement:
```yaml
kind: NodePool
apiVersion: karpenter.sh/v1
spec:
  template:
    metadata:
      labels:
        node.kubernetes.io/instance-type: 'c5.large'
```
I'd be curious how many people configure requirements for well-known labels this way, but given it's a valid configuration option we should support it (need to sanity check it is valid - I'm pretty sure it is but I've never tried it).

Reply:
I believe that it is valid -- though it is odd and I suspect that it's _significantly_ less common like you mentioned